src/Voron/Data/CompactTrees/PersistentDictionary.cs

```diff
@@ -0,0 +1,242 @@
+            
+            // Train the new dictionary.
+            var encoder = new HopeEncoder<Encoder3Gram<NativeMemoryEncoderState>>(
+                                new Encoder3Gram<NativeMemoryEncoderState>(
+                                    new NativeMemoryEncoderState(output.Ptr, UsableDictionarySize)));
+            encoder.Train(trainEnumerator, MaxDictionaryEntries);
+
+            // Test the new dictionary to ensure that we have statistically better compression.
+            using var encodeBufferScope = llt.Allocator.Allocate(Constants.Storage.PageSize, out var encodeBuffer);
+            
+            int incumbentSize = 0;
+            int successorSize = 0;
+            var auxEncodeBuffer = encodeBuffer.ToSpan();
+            while (testEnumerator.MoveNext(out var testValue))
+            {                
+                incumbentSize += previousDictionary._encoder.Encode(testValue, auxEncodeBuffer);
+                successorSize += encoder.Encode(testValue, auxEncodeBuffer);
+            }
+
+            // If the new dictionary is not at least 5% better, we return the current dictionary.             
+            if (incumbentSize < successorSize * 1.05)
+                return previousDictionary;            
+                                    
+            var p = llt.AllocatePage(NumberOfPagesForDictionary);
+            p.Flags = PageFlags.Overflow;
+            p.OverflowSize = NumberOfPagesForDictionary * Constants.Storage.PageSize;
+
+            PersistentDictionaryHeader* header = (PersistentDictionaryHeader*)p.DataPointer;
+            header->TableSize = UsableDictionarySize;
+            header->TableMaxEntries = MaxDictionaryEntries;
+            header->CurrentId = p.PageNumber;
+            header->PreviousId = previousDictionary != null ? previousDictionary.PageNumber : 0;
+            output.CopyTo(0, p.DataPointer + PersistentDictionaryHeader.SizeOf, 0, UsableDictionarySize);
+            header->TableHash = XXHash64.Calculate(p.DataPointer + sizeof(long), UsableDictionarySize + PersistentDictionaryHeader.SizeOf - sizeof(long));
+
+#if DEBUG
+            VerifyTable(p);
+#endif
+
+            return new PersistentDictionary(p);
+        }
+
+        public static void VerifyTable(Page page)
+        {
+            PersistentDictionaryHeader* header = (PersistentDictionaryHeader*)page.DataPointer;
+
+            // If checksum is not correct, we will throw a corruption exception
+            var tableHash = XXHash64.Calculate(page.DataPointer + sizeof(long), (ulong)header->TableSize + PersistentDictionaryHeader.SizeOf - sizeof(long));
+            if (tableHash != header->TableHash)
+                throw new VoronErrorException($"Persistent storage checksum mismatch, expected: {tableHash}, actual: {header->TableHash}");
+
+            // TODO: Perform more advanced encoding/decoding verifications to ensure the table is not corrupted. 
```

Comment:
Still TODO? If then then do we want to address it now or at later stage?

Reply:
Later stage, I forgot about that one. I will open a issue and remove it.